util/errorutil: centralize panic-catching logic

yuzefovich opened the pull request:
**sql/stats: don't ignore query error when retrieving stats protos for backup**

This commit fixes a minor bug introduced in 0fdb44abdc6cab54eb94c4d03651fe3515cdcfaa where we would swallow the query error if it was returned while iterating over query results in `getTableStatsProtosFromDB`. That method i inspired by `getTableStatsFromDB` (which returns an error in this case correctly), and both have a special case where decoding errors should be ignored but query errors shouldn't. To make things clearer additionally this commit renames some error variables.

**util/errorutil: centralize panic-catching logic**

Previously, we duplicated the same recover-in-defer logic in many places to prevent process from crashing by catching "safe" panics. This commit exposes the helper method instead. There should be no behavior change.

Note that we introduced an optional function callback argument, and it should be inlined (as well as the catcher itself), so there should be performance impact either.

Epic: None
Release note: None

yuzefovich committed:
util/errorutil: centralize panic-catching logic

Previously, we duplicated the same recover-in-defer logic in many places
to prevent process from crashing by catching "safe" panics. This commit
exposes the helper method instead. There should be no behavior change.

Note that we introduced an optional function callback argument, and it
should be inlined (as well as the catcher itself), so there should be
performance impact either.

Release note: None

jeffswenson reviewed (approved):
LGTM

michae2 reviewed (changes requested):
Thanks for doing this!

> @michae2 reviewed 11 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-:-Of0_352C5zE1jCk5-KL:bxgavm9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @mgartner)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0_DnL4YPXmJP2S6R3:-Of0_DnL4YPXmJP2S6R4:b5ggktq):*
I'm not sure I agree about the logging. Is there a way to keep those?
___
*[`pkg/util/errorutil/catch.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZSpfA_Fn2_Oh8bNc:-Of0ZSpfA_Fn2_Oh8bNd:b-ql8hh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4574e33db768c8aac92b1cd9367e8c78ad969e2/pkg/util/errorutil/catch.go#L38)):*
> ```Go
> 	if r := recover(); r != nil {
> 		if ok, e := shouldCatch(r); ok {
> 			*retErr = e
> ```

proposal: if retErr is already set (i.e. if we've panicked within an error path) it doesn't seem great to overwrite the existing error. WDYT of using `errors.CombineErrors` here?
___
*[`pkg/sql/opt/xform/optimizer.go` line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0Yo0Q8wVmwoOVWpMj:-Of0Yo0Q8wVmwoOVWpMk:b-s8bn0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2762008257a6ce0a4a183aa8b41f0a2a4f926ed/pkg/sql/opt/xform/optimizer.go#L261)):*
> ```Go
> 			if ok, e := errorutil.ShouldCatch(r); ok {
> 				err = e
> 				log.VEventf(o.ctx, 1, "%v", err)
> ```

Would be nice to add back this logging.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 1274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZF-57V64edaxspkZ:-Of0ZF-57V64edaxspk_:birvxit) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2762008257a6ce0a4a183aa8b41f0a2a4f926ed/pkg/sql/opt/exec/execbuilder/relational.go#L1274)):*
> ```Go
> 				if ok, e := errorutil.ShouldCatch(r); ok {
> 					err = e
> 					log.VEventf(ctx, 1, "%v", err)
> ```

Would be nice to bring back this logging.
___
*[`pkg/sql/plan_opt.go` line 1048 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0Z5rwCa3eFayRX60K:-Of0Z5rwCa3eFayRX60L:birvxit) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2762008257a6ce0a4a183aa8b41f0a2a4f926ed/pkg/sql/plan_opt.go#L1048)):*
> ```Go
> 			if ok, e := errorutil.ShouldCatch(r); ok {
> 				err = e
> 				log.VEventf(ctx, 1, "%v", err)
> ```

Would be nice to bring back this logging.
___
*[`pkg/sql/opt/exec/execbuilder/scalar.go` line 1198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZIid7IxoUe0UvCm4:-Of0ZIid7IxoUe0UvCm5:b-ttzfba) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2762008257a6ce0a4a183aa8b41f0a2a4f926ed/pkg/sql/opt/exec/execbuilder/scalar.go#L1198)):*
> ```Go
> 				if ok, e := errorutil.ShouldCatch(r); ok {
> 					err = e
> 					log.VEventf(ctx, 1, "%v", err)
> ```

logging
___
*[`pkg/sql/opt/exec/execbuilder/scalar.go` line 1409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZK295c8DxLGsa2dv:-Of0ZK295c8DxLGsa2dw:b-ttzfba) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2762008257a6ce0a4a183aa8b41f0a2a4f926ed/pkg/sql/opt/exec/execbuilder/scalar.go#L1409)):*
> ```Go
> 				if ok, e := errorutil.ShouldCatch(r); ok {
> 					err = e
> 					log.VEventf(ctx, 1, "%v", err)
> ```

logging


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql/stats: don't ignore query error when retrieving stats protos for backup

This commit fixes a minor bug introduced in
0fdb44abdc6cab54eb94c4d03651fe3515cdcfaa where we would swallow the
query error if it was returned while iterating over query results in
`getTableStatsProtosFromDB`. That method is inspired by
`getTableStatsFromDB` (which returns an error in this case correctly),
and both have a special case where decoding errors should be ignored but
query errors shouldn't. To make things clearer additionally this commit
renames some error variables.

Release note: None

yuzefovich commented on pkg/sql/stats/stats_cache.go:
```diff
@@ -1079,31 +1050,19 @@ func getTableStatsProtosFromDB(
 		}
 		statsProtos = append(statsProtos, tsp)
 	}
+	if err != nil {
```
Note that we do still ignore the error if it happens during table stats decoding (I just pushed a change to make this more explicit rather than using for-loop-local `err` that shadows the one in outer scope). The error that is being returned here is the one returned on `it.Next()` call and I think it was a bug that was introduced in 0fdb44abdc6cab54eb94c4d03651fe3515cdcfaa - in `getTableStatsFromDB` (which inspired `getTableStatsProtosFromDB`) we do return such an error.

I extracted the bug fix into a separate commit.
<!-- Reviewable comment -Of1k5-4-Mv8oEFtE4Dx:b-u04uke -->
<!-- Sent from Reviewable.io -->

michae2 replied:
Gah, sorry, I didn't notice that it was shadowed! Thanks for clarifying.
<!-- Reviewable comment -Of377B_Cuq0JIrcoj7F:b-6gpf5s -->
<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-:-Of1siXT6055kNJ-Gatz:bz5mair)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @mgartner, and @michae2)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0_DnL4YPXmJP2S6R3:-Of1rTQq3lB8ME6uQddB:bjmg7pz):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm not sure I agree about the logging. Is there a way to keep those?
</blockquote></details>

Ack, brought back the logging where it was present. (It doesn't look like we've been very principled about where to log vs not to log - it probably depended on where the panic-catching code was copied from. Actually, I was probably the one who introduced the first VEvent logging in 367ccec1ce5b9e6ffc84a8e28e638f18269f0948.)
___
*[`pkg/util/errorutil/catch.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZSpfA_Fn2_Oh8bNc:-Of1mK_tBDf13ayVTVZq:bgtkgkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4574e33db768c8aac92b1cd9367e8c78ad969e2/pkg/util/errorutil/catch.go#L38)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

proposal: if retErr is already set (i.e. if we've panicked within an error path) it doesn't seem great to overwrite the existing error. WDYT of using `errors.CombineErrors` here?
</blockquote></details>

I don't know whether we need to do this - the previous code didn't have this behavior, and supposedly if we're panicking, then `retErr` must not be set yet. I'm neutral to this idea, but if you think it's worth it, I'll add that in (I don't see how it can hurt).


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158369)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @mgartner)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158369)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @mgartner, and @michae2)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0_DnL4YPXmJP2S6R3:-Of5VOAvC2Lk15bZGzre:b-whzc26):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack, brought back the logging where it was present. (It doesn't look like we've been very principled about where to log vs not to log - it probably depended on where the panic-catching code was copied from. Actually, I was probably the one who introduced the first VEvent logging in 367ccec1ce5b9e6ffc84a8e28e638f18269f0948.)
</blockquote></details>

Making `errorutil` explicitly depend on `util/log` broke some disallowed tests (via `util/pretty` usage since this package is transitively used by `sem/tree`), so I refactored the approach to have an optional callback function. Both the catcher and the VEventf callbacks are inlined, so there should be no performance impact.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job!

(I went ahead and merged https://github.com/cockroachdb/cockroach/pull/158096 which has one more spot. If you want to get that one, too, I don't mind; or I can get it after this merges.)

> @michae2 reviewed 20 of 20 files at r4, 20 of 20 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-:-OfRjjWc9FnuTW4o_jEG:b29xp39)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jeffswenson, @mgartner, and @yuzefovich)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0_DnL4YPXmJP2S6R3:-OfRah69BpDfjLrWSzY2:b-9755fn):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Making `errorutil` explicitly depend on `util/log` broke some disallowed tests (via `util/pretty` usage since this package is transitively used by `sem/tree`), so I refactored the approach to have an optional callback function. Both the catcher and the VEventf callbacks are inlined, so there should be no performance impact.
</blockquote></details>

Thank you!
___
*[`pkg/util/errorutil/catch.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZSpfA_Fn2_Oh8bNc:-OfRmN8CEzpq8xH-uUib:b-nt4r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4574e33db768c8aac92b1cd9367e8c78ad969e2/pkg/util/errorutil/catch.go#L38)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I don't know whether we need to do this - the previous code didn't have this behavior, and supposedly if we're panicking, then `retErr` must not be set yet. I'm neutral to this idea, but if you think it's worth it, I'll add that in (I don't see how it can hurt).
</blockquote></details>

I'm mostly advocating for this since MaybeCatchPanic is becoming a "library" function, and future code might also use it. Seems worthwhile to harden it as much as we can.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I needed to handle that new call site too to make the code compile :smile: 

TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-:-OfS6nsp4dUn8zM9Udqq:b-lazpm9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jeffswenson, @mgartner, and @michae2)

___
*[`pkg/util/errorutil/catch.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158369#-Of0ZSpfA_Fn2_Oh8bNc:-OfS5w617BDt41NTjJl7:b-lhdfkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4574e33db768c8aac92b1cd9367e8c78ad969e2/pkg/util/errorutil/catch.go#L38)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm mostly advocating for this since MaybeCatchPanic is becoming a "library" function, and future code might also use it. Seems worthwhile to harden it as much as we can.
</blockquote></details>

Sounds reasonable, done.


<!-- Sent from Reviewable.io -->
